File: 04_dynamic_programming/placing_parentheses/placing_parentheses.cpp
#include <iostream>
#include <cassert>
#include <string>
#include <vector>
#include <limits.h>
using namespace std;

long long eval(long long a, long long b, char op) {
	if (op == '*') {
		return a * b;
	} else if (op == '+') {
		return a + b;
	} else if (op == '-') {
		return a - b;
	} else {
		assert(0);
	}
}

long long get_maximum_value(const string &exp) {
	//write your code here
	int size=exp.size();
	int n=(size+1)/2;//operands
	vector<vector<int> > m(n+1,vector<int>(n+1,0));
	vector<vector<int> > M(n+1,vector<int>(n+1,0));
	for(int i=1;i<=n;i++)
	{
		m[i][i]=exp[2*(i-1)]-'0';
		M[i][i]=exp[2*(i-1)]-'0';
	}
	for(int s=1;s<n;s++)
	{	for(int i=1;i<=n-s;i++)
		{
			int j=i+s;
			long long min1 = INT_MAX;
			long long max1=INT_MIN;
			for(int k=i;k<j;k++)
			{

				long long a=eval(M[i][k],M[k+1][j],exp[2*k-1]);
				long long b=eval(M[i][k],m[k+1][j],exp[2*k-1]);
				long long c=eval(m[i][k],M[k+1][j],exp[2*k-1]);
				long long d=eval(m[i][k],m[k+1][j],exp[2*k-1]);
				min1=min(min1,min(a,min(b,min(c,d))));
				max1=max(max1,max(a,max(b,max(c,d))));
			}
			m[i][j]=min1;
			M[i][j]=max1;
			
		}
	}
	return M[1][n];
}


int main() {
	string s;
	cin >> s;
	cout << get_maximum_value(s) << '\n';
}

File: 04_dynamic_programming/placing_parentheses/placing_parentheses.py
#Task. Find the maximum value of an arithmetic expression by specifying the order of applying its arithmetic operations using additional parentheses.
# Uses python3
import sys
def evalt(a, b, op):
    if op == '+':
        return a + b
    elif op == '-':
        return a - b
    elif op == '*':
        return a * b
    else:
        assert False

def get_maximum_value(dataset):
    size = len(dataset)
    n = (size+1)//2
    m = [[0 for x in range(n+1)] for y in range(n+1)]
    M = [[0 for x in range(n+1)] for y in range(n+1)]
    for i in range(1,n+1):
        m[i][i]=int(dataset[2*(i-1)])
        M[i][i]=int(dataset[2*(i-1)])
    for s in range(1,n+1):
        for i in range(1,n-s+1):
            j=i+s
            min1=sys.maxsize
            max1=-sys.maxsize-1
            for k in range(1,j):
                a=evalt(M[i][k],M[k+1][j],dataset[2*k-1])
                
                b=evalt(M[i][k],m[k+1][j],dataset[2*k-1])
                
                c=evalt(m[i][k],M[k+1][j],dataset[2*k-1])
                
                d=evalt(m[i][k],m[k+1][j],dataset[2*k-1])
                
                min1=min(min1,min(a,min(b,min(c,d))))
                max1=max(max1,max(a,max(b,max(c,d))))
            m[i][j]=min1
            M[i][j]=max1    
    return M[1][n]


if __name__ == "__main__":
    print(get_maximum_value(input()))
